tests/models/encoder_decoders/wikitables/wikitables_decoder_step_test.py

```diff
@@ -0,0 +1,346 @@
+                                            rnn_state=rnn_state,
+                                            grammar_state=grammar_state,
+                                            action_embeddings=self.action_embeddings,
+                                            action_indices=self.action_indices,
+                                            possible_actions=self.possible_actions,
+                                            flattened_linking_scores=flattened_linking_scores,
+                                            actions_to_entities=actions_to_entities,
+                                            entity_types=entity_types)
+
+    def test_get_actions_to_consider(self):
+        # pylint: disable=protected-access
+        valid_actions_1 = {'e': [0, 1, 2, 4]}
+        valid_actions_2 = {'e': [0, 1, 3]}
+        valid_actions_3 = {'e': [2, 3, 4]}
+        self.state.grammar_state[0] = GrammarState(['e'], {}, valid_actions_1, {})
+        self.state.grammar_state[1] = GrammarState(['e'], {}, valid_actions_2, {})
+        self.state.grammar_state[2] = GrammarState(['e'], {}, valid_actions_3, {})
+        self.state.action_embeddings = self.state.action_embeddings[:2]
+        considered, to_embed, to_link = WikiTablesDecoderStep._get_actions_to_consider(self.state)
+        # These are _global_ action indices.  They come from actions [[(0, 1), (0, 0)], [(1, 1)], []].
+        expected_to_embed = [[0, 1], [0], []]
+        assert to_embed == expected_to_embed
+        # These are _batch_ action indices with a _global_ action index above num_global_actions,
+        # sorted by their _global_ action index.
+        # They come from actions [[(0, 2), (0, 4)], [(1, 3), (1, 0)], [(0, 3), (0, 4)]].
+        expected_to_link = [[2, 4], [3, 0], [2, 3, 4]]
+        assert to_link == expected_to_link
+        # These are _batch_ action indices, sorted by _global_ action index, with padding in
+        # between the embedded actions and the linked actions.
+        expected_considered = [[1, 0, 2, 4, -1], [1, -1, 3, 0, -1], [-1, -1, 2, 3, 4]]
+        assert considered == expected_considered
+
+    def test_get_actions_to_consider_returns_none_if_no_linked_actions(self):
+        # pylint: disable=protected-access
+        valid_actions_1 = {'e': [0, 1, 2, 4]}
+        valid_actions_2 = {'e': [0, 1, 3]}
+        valid_actions_3 = {'e': [2, 3, 4]}
+        self.state.grammar_state[0] = GrammarState(['e'], {}, valid_actions_1, {})
+        self.state.grammar_state[1] = GrammarState(['e'], {}, valid_actions_2, {})
+        self.state.grammar_state[2] = GrammarState(['e'], {}, valid_actions_3, {})
+        considered, to_embed, to_link = WikiTablesDecoderStep._get_actions_to_consider(self.state)
+        # These are _global_ action indices.  All of the actions in this case are embedded, so this
+        # is just a mapping from the valid actions above to their global ids.
+        expected_to_embed = [[0, 1, 2, 5], [0, 3, 4], [2, 3, 5]]
```

Comment:
Why is the first element not `[1, 0, 2, 5]`, given that `self.action_indices[(0, 0)] = 1` and `self.action_indices[(0, 1)] = 0`?

Reply:
Because they are sorted by global action index.  This sorting is removed in the next PR, and the PR after that will probably rework how this works entirely, so I'm not going to worry about changing anything here.